mm2src/coins/z_coin/z_coin_errors.rs

```diff
@@ -1,258 +1,311 @@
+    }
 
-impl From<SqliteError> for GetUnspentWitnessErr {
-    fn from(err: SqliteError) -> GetUnspentWitnessErr { GetUnspentWitnessErr::Sql(err) }
-}
+    impl From<UtxoRpcError> for ZCoinBuildError {
+        fn from(err: UtxoRpcError) -> ZCoinBuildError { ZCoinBuildError::Rpc(err) }
+    }
+
+
+    impl From<std::io::Error> for ZCoinBuildError {
+        fn from(err: std::io::Error) -> ZCoinBuildError { ZCoinBuildError::Io(err) }
+    }
+
+
+
+    pub(super) enum SqlTxHistoryError {
+        Sql(SqlError),
+        FromIdDoesNotExist(i64),
+    }
+
+    impl From<SqlError> for SqlTxHistoryError {
+        fn from(err: SqlError) -> Self { SqlTxHistoryError::Sql(err) }
+    }
+
+    impl From<SqlTxHistoryError> for MyTxHistoryErrorV2 {
+        fn from(err: SqlTxHistoryError) -> Self {
+            match err {
+                SqlTxHistoryError::Sql(sql) => MyTxHistoryErrorV2::StorageError(sql.to_string()),
+                SqlTxHistoryError::FromIdDoesNotExist(id) => {
+                    MyTxHistoryErrorV2::StorageError(format!("from_id {} does not exist", id))
+                },
+            }
+        }
+    }
+
+    impl From<NoInfoAboutTx> for MyTxHistoryErrorV2 {
+        fn from(err: NoInfoAboutTx) -> Self {
+            MyTxHistoryErrorV2::RpcError(format!("No info about transaction {:02x}", err.0))
+        }
+    }
+
+    pub(super) struct NoInfoAboutTx(pub(super) H256Json);
+);
 
 #[derive(Debug, Display)]
-pub enum ZCoinBuildError {
-    UtxoBuilderError(UtxoCoinBuildError),
-    GetAddressError,
-    SqliteError(SqliteError),
-    Rpc(UtxoRpcError),
-    #[display(fmt = "Sapling cache DB does not exist at {}. Please download it.", path)]
-    SaplingCacheDbDoesNotExist {
-        path: String,
-    },
-    Io(std::io::Error),
-    RpcClientInitErr(ZcoinClientInitError),
-    ZCashParamsNotFound,
-    ZDerivationPathNotSet,
-    SaplingParamsInvalidChecksum,
-}
+#[display(fmt = "Blockchain scan process stopped")]
+pub struct BlockchainScanStopped {}
 
-impl From<SqliteError> for ZCoinBuildError {
-    fn from(err: SqliteError) -> ZCoinBuildError { ZCoinBuildError::SqliteError(err) }
-}
+cfg_wasm32!(
+    #[derive(Debug, Display)]
+    #[non_exhaustive]
+    pub enum UpdateBlocksCacheErr {}
 
-impl From<UtxoRpcError> for ZCoinBuildError {
-    fn from(err: UtxoRpcError) -> ZCoinBuildError { ZCoinBuildError::Rpc(err) }
-}
+    #[derive(Debug, Display)]
+    #[non_exhaustive]
+    pub enum ZcoinClientInitError {
+        BlocksDbInitFailure(String)
+    }
+
+    #[derive(Debug, Display)]
+    pub enum UrlIterError {}
+
+    #[derive(Debug, Display)]
+    pub enum GenTxError {}
+
+    #[allow(clippy::large_enum_variant)]
+    #[derive(Debug)]
+    pub enum SendOutputsErr {
+        GenTxError(GenTxError),
+        NumConversion(NumConversError),
+        Rpc(UtxoRpcError),
+    }
+
+    #[derive(Debug, Display)]
+    pub enum GetUnspentWitnessErr {}
```

Comment:
This is also duplicated, please go through all those duplications and resolve them in this PR. I will continue the review after this much needed code cleaning :)

Reply:
Copy that!